core-blocks/file/edit.js

```diff
@@ -0,0 +1,252 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { getBlobByURL, revokeBlobURL } from '@wordpress/utils';
+import {
+	ClipboardButton,
+	IconButton,
+	Toolbar,
+	withNotices,
+} from '@wordpress/components';
+import { Component, compose, Fragment } from '@wordpress/element';
+import {
+	MediaUpload,
+	MediaPlaceholder,
+	BlockControls,
+	RichText,
+	editorMediaUpload,
+} from '@wordpress/editor';
+
+/**
+ * Internal dependencies
+ */
+import './editor.scss';
+import FileBlockInspector from './inspector';
+import FileBlockEditableLink from './editable-link';
+
+class FileEdit extends Component {
+	constructor() {
+		super( ...arguments );
+
+		const {
+			href,
+			showDownloadButton = true,
+			buttonText = __( 'Download' ),
+			id,
+		} = this.props.attributes;
+
+		this.onSelectFile = this.onSelectFile.bind( this );
+
+		// Initialize default values if undefined
+		this.props.setAttributes( {
+			showDownloadButton,
+			buttonText,
+		} );
+
+		// edit component has its own attributes in the state so it can be edited
+		// without setting the actual values outside of the edit UI
+		this.state = {
+			showPlaceholder: ! href,
+			href,
+			attachmentPage: undefined,
+			showCopyConfirmation: false,
+		};
+
+		if ( id !== undefined ) {
+			// Get Attachment Page URL
+			wp.apiRequest( {
+				path: `/wp/v2/media/${ id }`,
+				method: 'GET',
+			} )
+				.then(
+					( result ) => {
+						this.setState( { attachmentPage: result.link } );
+					},
+					() => {
+						this.setState( { attachmentPage: undefined } );
+					}
+				);
+		}
+	}
+
+	componentDidMount() {
+		const { href } = this.state;
+
+		// Upload a file drag-and-dropped into the editor
+		if ( this.isBlobURL( href ) ) {
+			getBlobByURL( href )
+				.then( ( file ) => {
+					editorMediaUpload( {
+						allowedType: '*',
+						filesList: [ file ],
+						onFileChange: ( [ media ] ) => this.onSelectFile( media ),
+					} );
+					revokeBlobURL( href );
+				} );
+		}
+	}
+
+	componentDidUpdate( prevProps ) {
+		// Reset copy confirmation state when block is deselected
+		if ( prevProps.isSelected && ! this.props.isSelected ) {
+			this.setState( { showCopyConfirmation: false } );
+		}
+	}
+
+	onSelectFile( media ) {
+		if ( media && media.url ) {
+			// sets the block's attributes and updates the edit component from the
+			// selected media, then switches off the placeholder UI
+			this.props.setAttributes( {
+				href: media.url,
+				fileName: media.title,
+				textLinkHref: media.url,
+				id: media.id,
+			} );
+			this.setState( {
+				href: media.url,
+				fileName: media.title,
+				attachmentPage: media.link,
+				showPlaceholder: false,
+			} );
+		}
+	}
+
+	isBlobURL( url = '' ) {
```

Comment:
Would it make sense to move this one to `utils`, like the other blob functions? At which point would you split out `utils/blob`?

Reply:
Yes, I think so. But in a separate PR, because I think I saw a couple of other places in the codebase where we’re doing blob url detection, and I’d want to change them all at once.